File: internal/tool/current_time.go
package tool

import (
	"context"
	"encoding/json"
	"fmt"
	"time"
)

type CurrentTimeTool struct{}

func NewCurrentTimeTool() *CurrentTimeTool { return &CurrentTimeTool{} }

func (t *CurrentTimeTool) Name() string { return "current_time" }

func (t *CurrentTimeTool) Description() string {
	return "Get the current date and time. Use this when you need to know today's date, current time, day of week, or resolve relative time references like 'today', 'tomorrow', 'last week'."
}

func (t *CurrentTimeTool) Parameters() json.RawMessage {
	return json.RawMessage(`{
		"type": "object",
		"properties": {
			"timezone": {"type": "string", "description": "IANA timezone (e.g. America/New_York). Default: system local timezone."}
		}
	}`)
}

type currentTimeArgs struct {
	Timezone string `json:"timezone"`
}

func (t *CurrentTimeTool) Execute(_ context.Context, arguments string) (string, error) {
	var args currentTimeArgs
	json.Unmarshal([]byte(arguments), &args)

	loc := time.Now().Location()
	if args.Timezone != "" {
		parsed, err := time.LoadLocation(args.Timezone)
		if err != nil {
			return "", fmt.Errorf("invalid timezone %q: %w", args.Timezone, err)
		}
		loc = parsed
	}

	now := time.Now().In(loc)
	zone, offset := now.Zone()
	offsetHours := offset / 3600

	return fmt.Sprintf(
		"ISO 8601: %s\nDate: %s\nTime: %s\nDay: %s\nTimezone: %s (UTC%+d)\nUnix: %d\nWeek: %d, Day of year: %d",
		now.Format(time.RFC3339),
		now.Format("2006-01-02"),
		now.Format("15:04:05"),
		now.Format("Monday"),
		zone, offsetHours,
		now.Unix(),
		now.YearDay()/7+1, now.YearDay(),
	), nil
}

File: src/agent/tool/current_time.rs
use chrono::Utc;
use serde::Deserialize;
use serde_json::Value;

use super::Tool;

pub(super) struct CurrentTime;

#[derive(Deserialize)]
struct Args {
    timezone: Option<String>,
}

#[async_trait::async_trait]
impl Tool for CurrentTime {
    fn name(&self) -> &str {
        "current_time"
    }

    fn description(&self) -> &str {
        "Get the current date and time. Optionally specify a timezone (e.g. \"Asia/Shanghai\", \
         \"America/New_York\"). Defaults to UTC."
    }

    fn parameters(&self) -> Value {
        serde_json::json!({
            "type": "object",
            "properties": {
                "timezone": {
                    "type": "string",
                    "description": "IANA timezone name (e.g. Asia/Shanghai). Defaults to UTC."
                }
            },
            "required": []
        })
    }

    async fn execute(&self, args: &str) -> String {
        let parsed: Args = match serde_json::from_str(args) {
            Ok(a) => a,
            Err(e) => return format!("error: invalid arguments: {e}"),
        };

        let now = Utc::now();

        if let Some(ref tz_name) = parsed.timezone {
            let tz: chrono_tz::Tz = match tz_name.parse() {
                Ok(t) => t,
                Err(e) => return format!("error: invalid timezone \"{tz_name}\": {e}"),
            };
            let local = now.with_timezone(&tz);
            format!(
                "datetime: {}\ndate: {}\nday: {}\ntimezone: {}\nunix: {}",
                local.format("%Y-%m-%dT%H:%M:%S%:z"),
                local.format("%Y-%m-%d"),
                local.format("%A"),
                tz_name,
                now.timestamp(),
            )
        } else {
            format!(
                "datetime: {}\ndate: {}\nday: {}\ntimezone: UTC\nunix: {}",
                now.format("%Y-%m-%dT%H:%M:%S+00:00"),
                now.format("%Y-%m-%d"),
                now.format("%A"),
                now.timestamp(),
            )
        }
    }

    fn status_label(&self, _args: &str) -> String {
        "🕐 Checking time".into()
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[tokio::test]
    async fn valid_timezone() {
        let tool = CurrentTime;
        let result = tool.execute(r#"{"timezone": "Asia/Shanghai"}"#).await;
        assert!(
            result.contains("timezone: Asia/Shanghai"),
            "expected timezone in output, got: {result}"
        );
        assert!(result.contains("unix:"));
    }

    #[tokio::test]
    async fn invalid_timezone() {
        let tool = CurrentTime;
        let result = tool.execute(r#"{"timezone": "Foo/Bar"}"#).await;
        assert!(
            result.starts_with("error:"),
            "expected error, got: {result}"
        );
    }

    #[tokio::test]
    async fn empty_args_defaults_to_utc() {
        let tool = CurrentTime;
        let result = tool.execute("{}").await;
        assert!(
            result.contains("timezone: UTC"),
            "expected UTC default, got: {result}"
        );
    }

    #[tokio::test]
    async fn default_contains_unix() {
        let tool = CurrentTime;
        let result = tool.execute("{}").await;
        assert!(
            result.contains("unix:"),
            "expected unix timestamp, got: {result}"
        );
    }
}
